google-authenticator.php

```diff
@@ -237,21 +593,70 @@ function check_otp( $user, $username = '', $password = '' ) {
 					return new WP_Error( 'invalid_google_authenticator_password', __( '<strong>ERROR</strong>: The Google Authenticator password is incorrect.', 'google-authenticator' ) );
 				} 		 
 			} else {
-				return new WP_Error( 'invalid_google_authenticator_token', __( '<strong>ERROR</strong>: The Google Authenticator code is incorrect or has expired.', 'google-authenticator' ) );
-			}	
+				if ( ! $this->is_two_screen_signin_enabled() ) {
+					return new WP_Error( 'invalid_google_authenticator_token', __( '<strong>ERROR</strong>: The Google Authenticator code is incorrect or has expired.', 'google-authenticator' ) );
+				} else {
+					wp_logout();
+					$this->secondary_login_screen();
+					exit;
+				}
+			}
 		}
 	}
 	// Google Authenticator isn't enabled for this account,
 	// just resume normal authentication.
 	return $userstate;
 }
 
+function secondary_login_screen( $redirect_to ) {
+	$redirect_to = isset( $_REQUEST['redirect_to'] ) ? $_REQUEST['redirect_to'] : admin_url();
+	login_header( 'Secondary Login Screen' );
+	if ( array_key_exists( 'googleotp', $_REQUEST ) ) {
+		if ( 0 === strlen( $_REQUEST[ 'googleotp'] ) ) {
+			$error_message = __( '<strong>ERROR</strong>: The Google Authenticator code is missing.', 'google-authenticator' );
+		} else {
+			$error_message = __( '<strong>ERROR</strong>: The Google Authenticator code is incorrect or has expired.', 'google-authenticator' );
+		}
+		echo '<div id="login_error">' . $error_message . '</div>';
```

Comment:
which would you suggest given that <strong> tag is in there and the variable is hardcoded to one of two vals in lines 616/618

Reply:
@ivankruchkoff I'd go with `esc_html__()` here. I can imagine an attack via `gettext` filter used in `translate` function via any buggy 3rd party plugin (I know that if there is an option to inject PHP code there then attackers would find better ways of using it instead of tampering this `$error_message` but I am still a fan of late escaping of everything :))

Reply:
esc_html will escape the < strong > tag

Reply:
Oops my bad 👍 I missed this totally